thunks.go

```diff
@@ -192,12 +202,33 @@ type closure struct {
 	// arguments should be added to it, before executing it
 	env      environment
 	function *ast.Function
+	params   Parameters
 }
 
 func (closure *closure) EvalCall(arguments callArguments, e *evaluator) (value, error) {
 	argThunks := make(bindingFrame)
 	for i, arg := range arguments.positional {
-		argThunks[closure.function.Parameters.Positional[i]] = arg
+		if i < len(closure.function.Parameters.Required) {
+			argThunks[closure.function.Parameters.Required[i]] = arg
+		} else {
+			argThunks[closure.function.Parameters.Optional[i-len(closure.function.Parameters.Required)].Name] = arg
+		}
+
+	}
+
+	for _, arg := range arguments.named {
+		argThunks[arg.name] = arg.pv
+	}
+
+	defaultArgEnv := makeEnvironment(argThunks, selfBinding{})
```

Comment:
shouldn't they use calledEnvironment?

Reply:
They currently hold the captured environment inside, so the environment in which they are executed is correct. That said, maybe it can be simplified by using calledEnvironment. I'll look into this.